### DISCUSSION TITLE
How to get id of saved record in custom function?

### DISCUSSION THREAD

**Discussion by iamdineshkumar** [pocketbase/pocketbase#2114] [Q&A]
I am trying to save a record in OnRecordBeforeCreateRequest event hook and also need to save record id in relation field. How to access record id? whether record.id will return generated id after save?
```
				record := models.NewRecord(collection)
				record.Set("skill","Speach");
				if err := app.Dao().SaveRecord(record); err != nil {
					return err
				}
```

**Comment by ganigeorgiev** [comment]
> How to access record id? whether record.id will return generated id after save?

You can access the `record.Id` after the `SaveRecord()` call. Alternatively, you can also set a manually a custom Id upfront (or generate a random one with `record.RefreshId()`).

**Reply by iamdineshkumar** [reply]
I would like to give you context on what I am trying to do.
I have two tables Skills & Test
Skill Structure is just with skill field alone
Test have field skills which is relation with skill type and multiple skills can be stored in that field.
What I am trying to do is write a OnRecordBeforeCreateRequest event hook for test, before creating record store all skill values in db and replace their id in field.
```
collection, err := app.Dao().FindCollectionByNameOrId("skills")
if err != nil {
	return err
}
skillList:= struct {skills []string}{}
e.Record.UnmarshalJSONField("skills", &saleItems)
skillsId := []string{}
for _, item := range skillList.skillss {
	record := models.NewRecord(collection)
	record.Set("name", item)
	if err := app.Dao().SaveRecord(record); err != nil {
	return err
	}
	saleItemsId = append(saleItemsId, record.Id)
}
e.Record.Set("skills", skillsId )
```

Whether this code works?
If this is possible same way is that possible to send field value an array of json with differnt fields so that in similar way create relational records and assign thier id in field

Kindly guide me how to procced. In single request i am trying to handle this. Else if there is 15 skills then need to add one by one and then should collect thier id's to send final request of test. In future there may be requirements to store skill as well as category in skills table.

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand what the above code is supposed to do.

What is the request payload you are sending and want to process?

How is `skillList` populated?

`skillsId` doesn't seem to be changed anywhere.

Is `skills` a `json` or `relation` field? If it is a relation, then you can use directly the [`e.Record.GetStringSlice("skills")` helper](https://pocketbase.io/docs/record-methods/#getset-record-fields). But again, `saleItems` also doesn't seem to be used anywhere.

Please provide a more complete code sample.

**Reply by iamdineshkumar** [reply]
I will give you a little detailed information 
**Table Structure:**
_Table:_ skills
_Fields:_
id  name(String) category(String)

_Table:_ Test
_Fields:_
id 
testName (String)
skills(relational field for Skill table with multiple values)
testDate (DateTime)

Request Payload from sdk
```
{
    "testName":"Junior Assessment",
    "skills":[
        {
            "name":"ForeHand",
            "category":"Beginner"
        },
        {
            "name":"BackHand",
            "category":"Beginner"
        }
    ],
    "testDate":"2023-04-01"
}
```
In above payload, skills have value which are need to be inserted in skill before creating test. So I am using OnRecordBeforeCreateRequest event hook for test table. In which I would like to have below code
```
// Skill Struct
type skill struct {
	Name  string    `json:"name" db:"name"`
	Category string `json:"category" db:"category"`
}
collection, err := app.Dao().FindCollectionByNameOrId("skills")
if err != nil {
	return err
}
// List of Skill based on declared struct
skillList:= struct {skills []skill }{}

// Mapping List of Skill from payload
e.Record.UnmarshalJSONField("skills", &skillList)
// List of Skill ID for replacing it latter against existing json value
skillsId := []string{}
//itreating skill list to save all records in skill table
for _, item := range skillList.skillss {
	record := models.NewRecord(collection)
	record.Set("name", item.Name)
	record.Set("category", item.Category)
	if err := app.Dao().SaveRecord(record); err != nil {
	return err
	}
	// Inserted id is stored in temp variable
	skillsId = append(saleItemsId, record.Id)
}
// now we replaced actual payload with populated skillsid
e.Record.Set("skills", skillsId )
```

Sorry in above i have mentioned few variables wrongly.

Kindly let me know is the above is possible?

**Reply by ganigeorgiev** [reply]
If I'm understanding your use case, you can try something like (_note that I haven't tested it_):

```go
app.OnRecordBeforeCreateRequest("tests").Add(func(e *core.RecordCreateEvent) error {
    // read the submitted skills objects from the cache parsed request data
    // (alternatively, you can also read it directly with e.HttpContext.Bind(&skills) but keep in mind that the json body is a reader, meaning that it can be read only once)
    requestData := apis.RequestData(e.HttpContext)
    rawSkills, err := json.Marshal(requestData.Data["skills"])
    if err != nil {
        return apis.NewBadRequestError("Failed to read the submitted skills data", err)
    }

    skills := []struct {
        Name     string `json:"name"`
        Category string `json:"category"`
    }{}
    if err := json.Unmarshal(rawSkills, &skills); err != nil {
        return apis.NewBadRequestError("Failed to parse the submitted skills items", err)
    }

    skillsCollection, err := app.Dao().FindCollectionByNameOrId("skills")
    if err != nil {
        return apis.NewBadRequestError("Failed to fetch skills collection", err)
    }

    // iterate and create each skills item
    skillIds := make([]string, 0, len(skills))
    txErr := app.Dao().RunInTransaction(func(txDao *daos.Dao) error {
        for _, skill := range skills {
            model := models.NewRecord(skillsCollection)
            model.Set("name", skill.Name)
            model.Set("category", skill.Category)

            if err := txDao.SaveRecord(model); err != nil {
                return err
            }

            skillIds = append(skillIds, model.Id)
        }

        return nil
    })

    if txErr != nil {
        return apis.NewBadRequestError("Failed to create all submitted skills", txErr)
    }

    e.Record.Set("skills", skillIds)

    return nil
})
```

Some notes:

- Since the above is happening in a before request hook, it will not revert the newly created skills if for some reason the main model persistence fail.
- I'd recommend to use a different key than the original field name when you are submitting custom formatted values (eg. instead of `skills` in the payload you can use `skillsList` or similar to differentiate from the regular skills format submission).
